base/strings/io.jl

```diff
@@ -354,12 +354,16 @@ function join(io::IO, iterator, delim="")
 end
 
 function _join_preserve_annotations(iterator, args...)
-    if _isannotated(eltype(iterator)) || any(_isannotated, args)
+    if isconcretetype(eltype(iterator)) && !_isannotated(eltype(iterator)) && !any(_isannotated, args)
+        sprint(join, iterator, args...)
+    else
         io = AnnotatedIOBuffer()
         join(io, iterator, args...)
-        read(seekstart(io), AnnotatedString{String})
-    else
-        sprint(join, iterator, args...)
+        if isconcretetype(eltype(iterator)) || !isempty(io.annotations)
+            read(seekstart(io), AnnotatedString{String})
+        else
+            String(take!(io.io))
```

Comment:
So this line is the essential new bit, right (the rest is moving around and tweaking logic to ensure this one is called in a specific case).

Just from looking at this, I'd wonder "what the hell is going on here, why can't we always use the `read` (or always the `take!`) approach?

Perhaps you could add a comment here briefly explaining the rationale?

Reply:
Sure! How does this look to you?

```
# If we know (from compile time information, or dynamically in the case
# of iterators with a non-concrete eltype), that the result is annotated
# in nature, we extract an `AnnotatedString`, otherwise we just extract
# a plain `String` from `io`.
```

Reply:
I've just pushed the above.